### PR TITLE
Short-circuit ToArray method if source is an array

### PR DESCRIPTION
**Pull request by alrz**

### CODE REVIEW

File: src/System.Linq/src/System/Linq/ToCollection.cs [hunk 1]
```diff
@@ -14,9 +14,18 @@ public static TSource[] ToArray<TSource>(this IEnumerable<TSource> source)
             {
                 throw Error.ArgumentNull(nameof(source));
             }
+            
+            if (source is TSource[] array)
+            {
+                return array;
+            }
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/ToCollection.cs]: @VSaDov can correct me if I'm wrong, but I believe we explicitly decided against this as ToArray is meant to have copy semantics, and returning the same array will allow writes to the returned array to be visible on the original.

**Comment by alrz** [comment]
Would it be desirable to use `Array.Copy` in `EnumerableHelpers.ToArray` specifically for arrays?

**Comment by stephentoub** [comment]
That would need to be (partially) answered by perf tests. If you do such tests and can show that such a change makes a measurable improvement for arrays of various sizes, then we'd need to understand what the impact is on other types, and determine whether it's a good trade-off based on expected scenarios and usage.

**Comment by alrz** [comment]
Array is an ICollection, and wrt copy semantics this doesn't need to be changed.

File: src/System.Linq/src/System/Linq/ToCollection.cs (hunk 1 above)
Comment (VSadov) [inline src/System.Linq/src/System/Linq/ToCollection.cs]: Right. FWIW the change can easily break code relying on existing shallow-copy behavior.

**Pull request closed without merging**